socs/agents/camera_rtsp/agent.py

```diff
@@ -0,0 +1,887 @@
+
+        # Create the recording buffer on disk
+        self.vid_buffer = CircularMediaBuffer(
+            self.vid_dir,
+            "vid",
+            "mp4",
+            max_record_files,
+            video_write_callback,
+            read_callback=None,
+            recent=0,
+            fps=record_fps,
+        )
+
+        # Register OCS feed
+        self.feed_name = f"RTSP_{self.address}"
+
+        agg_params = {"frame_length": self.seconds}  # [sec]
+        self.agent.register_feed(
+            self.feed_name,
+            record=True,
+            agg_params=agg_params,
+            buffer_time=1.0,
+        )
+
+    @ocs_agent.param("test_mode", default=False, type=bool)
+    def acq(self, session, params=None):
+        """acq(test_mode=False)
+
+        **Process** - Capture frames from an RTSP camera.
+
+        Args:
+            test_mode (bool):  Run the Process loop only once. Meant only
+                for testing.  Default is False.
+
+        Notes:
+            Individual frames are written to a circular disk buffer.  Metadata
+            about the captured images is stored in the session data.  The format
+            of this is::
+
+                >>> response.session['data']
+                {
+                    'address': 'camera-c1.simonsobs.org',
+                    'timestamp': 1701983575.123456,
+                    'path': '/ocs/cameras_rtsp/c1/img_2023-12-29T02:44:47+00:00.jpg',
+                }
+
+        """
+        pm = Pacemaker(1 / self.seconds, quantize=False)
+        pmgrab = Pacemaker(self.record_fps, quantize=True)
+
+        frames_per_snapshot = int(self.seconds * self.record_fps)
+
+        session.set_status("running")
+        self.is_streaming = True
+
+        # Open camera stream
+        if self.fake:
+            cap = FakeCamera()
+        else:
+            cap = cv2.VideoCapture(self.connection)
+        if not cap.isOpened():
+            self.log.error(f"Cannot open RTSP stream at {self.connection}")
+            return False, "Could not open RTSP stream"
+
+        # Tracking state of whether we are currently recording motion detection
+        detecting = False
+        detect_start = None
+        record_frames = int(self.record_fps * self.record_duration)
+        motion_detector = MotionDetector()
+
+        snap_count = 0
+        while self.is_streaming:
+            # Use UTC
+            timestamp = time.time()
+            data = dict()
+
+            for iframe in range(frames_per_snapshot):
+                pmgrab.sleep()
+                # Grab an image
+                _ = cap.grab()
+            success, image = cap.retrieve()
+            if not success:
+                msg = "Failed to retrieve snapshot image from stream"
+                self.log.error(msg)
+                return False, "Broken stream"
+
+            # Motion detection.  We ignore the first few snapshots and also
+            # any changes that happen while we are already recording.
```

Comment:
Is there any way to increase the recording duration if motion is continuously being detected? I figure we'd want to have a continuous video of a break-in that occurs longer than 60 seconds rather than having to open many 60 second videos.

Reply:
I'll see how difficult it is to change the logic.  In the worst case, we could increase the recording time to something longer and ensure that it is only enabled at night.

Reply:
This is quite challenging to implement with the current logic and I am not going to do this now.  Hopefully it is sufficient to increase the recording time to a few minutes and only enable motion detection during the night.